### ISSUE TITLE
About the role of AEnet

### ISSUE THREAD

**Issue by jianlong-yuan** [heyufan1995/self-domain-adapted-network#2]
Hi, excellent work. But I have some questions about AEnet:
I found that AEnet was not used in the testing phase. What is the function of all AEnet?

![image](https://user-images.githubusercontent.com/18137551/126469859-c83858b3-0493-46ae-868d-2c1c315362cc.png)

![image](https://user-images.githubusercontent.com/18137551/126469902-a8998b51-8b14-4f2f-989d-fa10e53d1075.png)

**Comment by YufanHe** [comment]
Hi:
the adaptation using AEs is not written in test.
please refer to train.py  model.opt_ANet
Thanks,
Yufan

On Jul 21, 2021, at 03:51, CodingMan ***@***.***> wrote:

﻿

      External Email - Use Caution



Hi, excellent work. But I have some questions about AEnet:
I found that AEnet was not used in the testing phase. What is the function of all AEnet?

def test(self, return_pred=False): """Test using ANet and TNet """ self.TNet.eval() self.ANet.eval() with torch.no_grad(): pred = self.ANet(self.image, self.TNet, self.AENet[0], side_out=True) pred, adapt_img = pred[-1], pred[0] pred_na = self.TNet(self.image) pred = F.softmax(pred, dim=1) pred_na = F.softmax(pred_na, dim=1) batch_size = self.image.shape[0] if self.opt.saveimage: for b_ix in range(batch_size): ids = os.path.split(self.filename[b_ix])[-1].split('.')[0] self.plot(tonp(adapt_img[b_ix]), ids + '_adapt.png') self.plot(tonp(pred[b_ix]), ids + '_preda.png') self.plot(tonp(self.label[b_ix]), ids + '_label.png') self.plot(tonp(pred_na[b_ix]), ids + '_predna.png') self.plot(tonp(self.image[b_ix]), ids + '_image.png') metric = [[],[]] if self.opt.__dict__.get('cal_metric',True): metric[0] = self.cal_metric(pred, self.label.squeeze(1)) metric[1] = self.cal_metric(pred_na, self.label.squeeze(1)) if return_pred: return metric, pred, pred_na else: return metric

def forward(self, x, TNet, AENet, side_out=False): """ Forward for a 4-level UNet Args: TNet: nn.Module. The pretrained task network side_out: bool. If true, output every intermediate results seq: list->int or np array. Position of 1x1 convolution """ x = self.adpNet(x) xh = [x] x = TNet.inblocks(x) ct = 0 # apply 1x1 conv on input blocks seq = self.seq if ct in seq: x = self.conv[ct](x) ct += 1 xh.append(x) for i in range(TNet.depth): x = TNet.downblocks[i](x) # apply 1x1 conv on every downsample output except bottleneck if ct in seq: x = self.conv[ct](x) ct += 1 xh.append(x) x = TNet.bottleneck(x) if ct in seq: x = self.conv[ct](x) ct += 1 xh.append(x) for i in range(TNet.depth): x = TNet.upblocks[TNet.depth-i-1](x,xh[TNet.depth-i]) if ct in seq: x = self.conv[ct](x) ct += 1 xh.append(x) x = TNet.outblock(x) xh.append(x) if side_out: return xh else: return x
